database/buildSchema/43_views_functions_triggers.sql

```diff
@@ -630,21 +630,64 @@ IMMUTABLE;
 ALTER TABLE review_assignment
     ADD COLUMN IF NOT EXISTS template_id INT GENERATED ALWAYS AS (review_assignment_template_id (application_id)) STORED;
 
-CREATE OR REPLACE FUNCTION public.empty_assigned_sections ()
+-- These no longer used as combined into (below) validation function
+DROP FUNCTION IF EXISTS public.empty_assigned_sections ();
+
+DROP TRIGGER IF EXISTS review_assignment_trigger2 ON public.review_assignment;
+
+-- Enforce validity of assigned sections:
+-- If status is now AVAILABLE assigned sections should be empty
+-- Also check that assigned sections aren't already assigned and remove if so
+-- This all happens *BEFORE* the record is inserted
+CREATE OR REPLACE FUNCTION public.enforce_asssigned_section_validity ()
     RETURNS TRIGGER
-    AS $review_assignment_event$
+    AS $trigger_queue$
 BEGIN
-    UPDATE
-        public.review_assignment
-    SET
-        assigned_sections = '{}'
-    WHERE
-        id = NEW.id;
-    RETURN NULL;
+    IF NEW.status = 'AVAILABLE' THEN
+        NEW.assigned_sections = '{}';
+    ELSE
+        NEW.assigned_sections = ARRAY ( WITH a AS (
+                SELECT
+                    unnest(NEW.assigned_sections) new_assigned
+                FROM
+                    review_assignment
+                WHERE
+                    id = NEW.id
+)
+                SELECT
+                    new_assigned
+                FROM
+                    a
+                WHERE
+                    new_assigned NOT IN (
+                        SELECT
+                            unnest(assigned_sections)
+                        FROM
+                            review_assignment
+                        WHERE
+                            stage_id = NEW.stage_id
+                            AND level_number = NEW.level_number
+                            AND reviewer_id <> NEW.reviewer_id)
+                        AND (new_assigned = ANY (NEW.allowed_sections)
+                            OR NEW.allowed_sections IS NULL));
+    END IF;
+    IF NEW.assigned_sections = '{}' THEN
+        NEW.status = 'AVAILABLE';
+    END IF;
+    NEW.time_updated = NOW();
+    RETURN NEW;
```

Comment:
This is the magic improvement: `RETURN NEW;` So much better than using `UPDATE` on the record directly as it completely eliminates the chance of an infinite loop.